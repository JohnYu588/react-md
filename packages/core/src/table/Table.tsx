import { cnb } from "cnbuilder";
import type { TableHTMLAttributes } from "react";
import { forwardRef, useMemo } from "react";
import { bem } from "../utils";
import type {
  TableConfigContext,
  TableConfiguration,
} from "./TableConfigurationProvider";
import { TableConfigProvider } from "./TableConfigurationProvider";

const styles = bem("rmd-table");

/** @remarks \@since 6.0.0 */
export interface TableClassNameOptions {
  className?: string;

  /** @defaultValue `false` */
  dense?: boolean;

  /** @defaultValue `false` */
  fullWidth?: boolean;
}

/**
 * @remarks \@since 6.0.0
 */
export function table(options: TableClassNameOptions = {}): string {
  const { dense = false, fullWidth = false, className } = options;

  return cnb(styles({ dense, "full-width": fullWidth }), className);
}

export interface TableProps
  extends TableHTMLAttributes<HTMLTableElement>,
    TableConfiguration {}

/**
 * @example
 * Responsive Example
 * ```tsx
 * import {
 *   Table,
 *   TableBody,
 *   TableCell,
 *   TableContainer,
 *   TableHeader,
 *   TableRow,
 * } from "@react-md/core";
 * import type { ReactElement } from "react";
 *
 * function Example(): ReactElement {
 *   return (
 *     <TableContainer>
 *       <Table>
 *         <TableHeader>
 *           <TableRow>
 *             <TableCell>Header 1</TableCell>
 *             <TableCell>Header 2</TableCell>
 *             <TableCell>Header 3</TableCell>
 *           </TableRow>
 *         </TableHeader>
 *         <TableBody>
 *           <TableRow>
 *             <TableCell>Row 1 Cell 1</TableCell>
 *             <TableCell>Row 1 Cell 2</TableCell>
 *             <TableCell>Row 1 Cell 3</TableCell>
 *           </TableRow>
 *           <TableRow>
 *             <TableCell>Row 2 Cell 1</TableCell>
 *             <TableCell>Row 2 Cell 2</TableCell>
 *             <TableCell>Row 2 Cell 3</TableCell>
 *           </TableRow>
 *           <TableRow>
 *             <TableCell>Row 3 Cell 1</TableCell>
 *             <TableCell>Row 3 Cell 2</TableCell>
 *             <TableCell>Row 3 Cell 3</TableCell>
 *           </TableRow>
 *         </TableBody>
 *       </Table>
 *     </TableContainer>
 *   );
 * }
 * ```
 */
export const Table = forwardRef<HTMLTableElement, TableProps>(function Table(
  props,
  ref
) {
  const {
    className,
    children,
    dense = false,
    hAlign = "left",
    vAlign = "middle",
    lineWrap = false,
    fullWidth = false,
    disableHover = false,
    disableBorders = false,
    ...remaining
  } = props;

  const configuration = useMemo<TableConfigContext>(
    () => ({
      dense,
      header: false,
      hAlign,
      vAlign,
      lineWrap,
      disableHover,
      disableBorders,
    }),
    [dense, hAlign, vAlign, lineWrap, disableHover, disableBorders]
  );

  return (
    <TableConfigProvider value={configuration}>
      <table
        {...remaining}
        ref={ref}
        className={table({ dense, fullWidth, className })}
      >
        {children}
      </table>
    </TableConfigProvider>
  );
});
