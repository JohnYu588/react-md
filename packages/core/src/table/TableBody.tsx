"use client";
import { cnb } from "cnbuilder";
import { forwardRef, useMemo, type HTMLAttributes } from "react";
import {
  TableConfigProvider,
  useTableConfig,
} from "./TableConfigurationProvider.js";
import { type TableConfig, type TableConfigContext } from "./types.js";

export interface TableBodyProps
  extends HTMLAttributes<HTMLTableSectionElement>,
    Omit<TableConfig, "header"> {}

/**
 * **Client Component**
 *
 * Creates a `<tbody>` element that also allows for overriding all the child
 * `TableCell` components with additional styling behavior.
 */
export const TableBody = forwardRef<HTMLTableSectionElement, TableBodyProps>(
  function TableBody(props, ref) {
    const {
      className,
      children,
      hAlign: propHAlign,
      vAlign: propVAlign,
      lineWrap: propLineWrap,
      disableHover: propDisableHover,
      disableBorders: propDisableBorders,
      ...remaining
    } = props;

    // update the table configuration with the custom overrides for the `<thead>`
    const { dense, hAlign, vAlign, lineWrap, disableHover, disableBorders } =
      useTableConfig({
        hAlign: propHAlign,
        vAlign: propVAlign,
        lineWrap: propLineWrap,
        disableHover: propDisableHover,
        disableBorders: propDisableBorders,
      });

    const configuration = useMemo<TableConfigContext>(
      () => ({
        header: false,
        dense,
        hAlign,
        vAlign,
        lineWrap,
        disableBorders,
        disableHover,
      }),
      [dense, hAlign, vAlign, lineWrap, disableBorders, disableHover]
    );

    return (
      <TableConfigProvider value={configuration}>
        <tbody {...remaining} ref={ref} className={cnb("rmd-tbody", className)}>
          {children}
        </tbody>
      </TableConfigProvider>
    );
  }
);
