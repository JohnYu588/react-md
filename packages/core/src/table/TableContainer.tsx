"use client";
import { cnb } from "cnbuilder";
import type { HTMLAttributes } from "react";
import { forwardRef, useMemo } from "react";
import { useEnsuredRef } from "../useEnsuredRef";
import type { TableContainerContext } from "./TableContainerProvider";
import { TableContainerProvider } from "./TableContainerProvider";

/**
 * @remarks \@since 6.0.0
 */
export function tableContainer({ className }: { className?: string }): string {
  return cnb("rmd-table-container", className);
}

export type TableContainerProps = HTMLAttributes<HTMLDivElement>;

/**
 * **Client Component**
 *
 * An extremely "useful" component that should be used with the `Table`
 * component if you want to make a responsive table within the page. If you
 * don't want to use this component, you can just apply `overflow: auto` to a
 * parent element of the table.
 */
export const TableContainer = forwardRef<HTMLDivElement, TableContainerProps>(
  function TableContainer(props, ref) {
    const { className, children, ...remaining } = props;
    const [nodeRef, refCallback] = useEnsuredRef(ref);

    const value = useMemo<TableContainerContext>(
      () => ({
        exists: true,
        containerRef: nodeRef,
      }),
      [nodeRef]
    );

    return (
      <TableContainerProvider value={value}>
        <div
          {...remaining}
          ref={refCallback}
          className={tableContainer({ className })}
        >
          {children}
        </div>
      </TableContainerProvider>
    );
  }
);
