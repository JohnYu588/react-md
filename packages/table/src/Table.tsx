import { bem } from "@react-md/core";
import { cnb } from "cnbuilder";
import type { TableHTMLAttributes } from "react";
import { forwardRef, useMemo } from "react";

import type {
  TableConfigContext,
  TableConfiguration,
} from "./TableConfigurationProvider";
import { TableConfigProvider } from "./TableConfigurationProvider";

const styles = bem("rmd-table");

export interface TableClassNameOptions {
  className?: string;
  dense?: boolean;
  fullWidth?: boolean;
}

export function table(options: TableClassNameOptions): string {
  const { dense, fullWidth, className } = options;

  return cnb(styles({ dense, "full-width": fullWidth }), className);
}

export interface TableProps
  extends TableHTMLAttributes<HTMLTableElement>,
    TableConfiguration {}

export const Table = forwardRef<HTMLTableElement, TableProps>(function Table(
  props,
  ref
) {
  const {
    className,
    children,
    dense = false,
    hAlign = "left",
    vAlign = "middle",
    lineWrap = false,
    fullWidth = false,
    disableHover = false,
    disableBorders = false,
    ...remaining
  } = props;

  const configuration = useMemo<TableConfigContext>(
    () => ({
      dense,
      header: false,
      hAlign,
      vAlign,
      lineWrap,
      disableHover,
      disableBorders,
    }),
    [dense, hAlign, vAlign, lineWrap, disableHover, disableBorders]
  );

  return (
    <TableConfigProvider value={configuration}>
      <table
        {...remaining}
        ref={ref}
        className={table({ dense, fullWidth, className })}
      >
        {children}
      </table>
    </TableConfigProvider>
  );
});
