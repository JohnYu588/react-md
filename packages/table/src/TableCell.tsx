import type { PropsWithRef } from "@react-md/core";
import { bem, useIcon } from "@react-md/core";
import { cnb } from "cnbuilder";
import type {
  ButtonHTMLAttributes,
  ReactNode,
  TdHTMLAttributes,
  ThHTMLAttributes,
} from "react";
import { forwardRef } from "react";

import type { SortOrder } from "./TableCellContent";
import { TableCellContent } from "./TableCellContent";
import type {
  TableCellConfig,
  TableCellHorizontalAlignment,
  TableCellVerticalAlignment,
} from "./TableConfigurationProvider";
import { useTableConfig } from "./TableConfigurationProvider";

export type TableCellAttributes = Omit<
  | TdHTMLAttributes<HTMLTableCellElement>
  | ThHTMLAttributes<HTMLTableCellElement>,
  "scope"
>;

/**
 * @remarks \@since 6.0.0
 */
export interface TableCellOptions extends TableCellConfig {
  /**
   * This is a bit of a "weird" prop since all it does is apply `width: 100%` to
   * this cell. This will make this specific cell fill the remaining width of
   * the table (if there is any). If no cells have this prop enabled and the
   * `fullWidth` table configuration is enabled, all cells will have an
   * equal-sized width.
   *
   * @defaultValue `false`
   */
  grow?: boolean;

  /**
   * This prop is only valid when the `header` prop is enabled or the
   * `TableCell` is a child of the `TableHeader` component. This will generally
   * be used with a value of `"row"` if you have table headers that are at the
   * start of each row instead of at the top of the table.
   *
   * @defaultValue `"col"`
   */
  scope?: "row" | "col" | "rowgroup" | "colgroup";

  /**
   *
   * @defaultValue `false`
   * @remarks \@since 6.0.0 This prop is only a boolean.
   */
  sticky?: boolean;

  /**
   * @internal
   * @defaultValue `false`
   */
  checkbox?: boolean;
}

/**
 * @remarks \@since 6.0.0 Removed the `colSpan="100%"` support since `colSpan`
 * really only supports numbers.
 */
export interface TableCellProps extends TableCellAttributes, TableCellOptions {
  /**
   * If you want to apply a sort icon for a header cell, set this prop to either
   * `"ascending"` or `"descending"`. When you change the sort order, this prop
   * should change as well which will cause the sort icon to rotate. The default
   * behavior is to have the icon facing upwards and not-rotated when
   * `"ascending"`, otherwise it will rotate downwards when `"descending"`.
   *
   * If this prop is set to `"none"`, the cell will render the clickable button
   * in the children, just without the sort icon. This is so that the sort
   * behavior can still be toggled for keyboard users and will be tab-focusable.
   */
  "aria-sort"?: SortOrder;

  /**
   * An optional sort icon to use. This will be defaulted to the configured sort
   * icon from the `@react-md/core` package. If you do not want the default
   * implementation for the sort icon behavior from `react-md`, you can set this
   * prop to `null`.
   */
  sortIcon?: ReactNode;

  /**
   * Boolean if the sort icon should appear after the children in the cell
   * instead of before.
   */
  sortIconAfter?: boolean;

  /**
   * Boolean if the sort icon should be rotated instead of the default
   * direction. When this is `undefined`, it will only be `true` when the
   * `"aria-sort"` prop is set to `"descending"`. If this is not `undefined`,
   * its boolean value will always be used.
   */
  sortIconRotated?: boolean;

  /**
   * Boolean if cell should no longer have any padding since you want a child
   * element to span the entire size of the cell instead. This is helpful when
   * rendering clickable and focusable elements within a cell.
   *
   * This will be defaulted to `true` if the `"aria-sort"` prop has been
   * provided and the `sortIcon` is not resoled as `null`. You can override this
   * default behavior by manually setting this to `true` or `false`.
   */
  disablePadding?: boolean;

  contentProps?: PropsWithRef<
    ButtonHTMLAttributes<HTMLButtonElement>,
    HTMLButtonElement
  >;
}

const styles = bem("rmd-table-cell");

/** @remarks \@since 6.0.0 */
export interface TableCellClassNameOptions {
  className?: string;

  /**
   * Set this to `true` if the cell is rendered as a `<th>` so that the correct
   * sticky styles can be applied.
   */
  header: boolean;

  /**
   * Set this to true if the cell is rendered in a `<thead>` so that the correct
   * sticky styles can be applied.
   */
  isInTableHeader: boolean;

  /** @defaultValue `false` */
  grow?: boolean;

  /** @defaultValue `false` */
  sticky?: boolean;

  /** @defaultValue `false` */
  checkbox?: boolean;

  hAlign?: TableCellHorizontalAlignment;
  vAlign?: TableCellVerticalAlignment;

  /** @defaultValue `true` */
  lineWrap?: boolean | "padded";

  /** @defaultValue `false` */
  disablePadding?: boolean;
}

/**
 * @remarks \@since 6.0.0
 */
export function tableCell(options: TableCellClassNameOptions): string {
  const {
    grow,
    sticky,
    header,
    checkbox,
    hAlign = "",
    vAlign = "",
    lineWrap = true,
    disablePadding,
    isInTableHeader,
    className,
  } = options;

  return cnb(
    styles({
      grow,
      header,
      sticky,
      checkbox,
      "sticky-cell": sticky && (!isInTableHeader || checkbox),
      "sticky-header": sticky && isInTableHeader,
      "sticky-header-cell": sticky && isInTableHeader && checkbox,
      [hAlign]: hAlign !== "left",
      [vAlign]: vAlign !== "middle",
      vertical: vAlign !== "middle",
      "no-wrap": !lineWrap,
      padded: !disablePadding && lineWrap === "padded",
      "no-padding": disablePadding,
    }),
    className
  );
}

/**
 * Creates a `<th>` or `<td>` cell with sensible styled defaults. You can create
 * a `<th>` element by enabling the `header` prop OR having a `TableCell` as a
 * child of the `TableHeader` component.
 *
 * Note: If you have a checkbox column in the `TableHeader` without any labels,
 * you will need to manually set the `header={false}` prop for that cell since
 * it is invalid to have a `<th>` without any readable content for screen
 * readers.
 */
export const TableCell = forwardRef<HTMLTableCellElement, TableCellProps>(
  function TableCell(props, ref) {
    const {
      "aria-sort": sortOrder,
      className,
      grow = false,
      scope: propScope,
      hAlign: propHAlign,
      vAlign: propVAlign,
      header: propHeader,
      lineWrap: propDisableLineWrap,
      checkbox,
      children,
      sticky,
      sortIcon: propSortIcon,
      sortIconAfter = false,
      sortIconRotated,
      disablePadding,
      contentProps,
      ...remaining
    } = props;

    const sortIcon = useIcon("sort", propSortIcon);
    const isNoPadding = disablePadding ?? !!(sortIcon && sortOrder);

    // Note: unlike the other usages of `useTableConfig`, the `propHeader`
    // is not provided. This is so that `TableCheckbox` components can still
    // be a sticky header without being rendered as a `<th>`. This also makes
    // it so the scope can be defaulted to `col` or `row` automatically.
    const {
      header: inheritedHeader,
      hAlign,
      vAlign,
      lineWrap,
    } = useTableConfig({
      hAlign: propHAlign,
      vAlign: propVAlign,
      lineWrap: propDisableLineWrap,
    });
    const header = propHeader ?? inheritedHeader;

    let scope = propScope;
    if (!scope && header) {
      scope = !inheritedHeader && propHeader ? "row" : "col";
    }

    const Component = header ? "th" : "td";
    return (
      <Component
        {...remaining}
        ref={ref}
        aria-sort={sortOrder === "none" ? undefined : sortOrder}
        className={tableCell({
          className,
          grow,
          header,
          sticky,
          checkbox,
          hAlign,
          vAlign,
          lineWrap: !sortOrder && lineWrap,
          disablePadding: isNoPadding,
          isInTableHeader: inheritedHeader,
        })}
        scope={scope}
      >
        <TableCellContent
          {...contentProps}
          icon={sortIcon}
          iconAfter={sortIconAfter}
          sortOrder={sortOrder}
          hAlign={hAlign}
          rotated={sortIconRotated}
        >
          {children}
        </TableCellContent>
      </Component>
    );
  }
);
