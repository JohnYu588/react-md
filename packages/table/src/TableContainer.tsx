import { useEnsuredRef } from "@react-md/core";
import { cnb } from "cnbuilder";
import type { HTMLAttributes, RefObject } from "react";
import { createContext, forwardRef, useContext, useMemo } from "react";

/**
 * @internal
 * @remarks \@since 6.0.0
 */
export interface TableContainerContext {
  exists: boolean;
  containerRef: RefObject<HTMLDivElement>;
}

const context = createContext<Readonly<TableContainerContext>>({
  exists: false,
  containerRef: { current: null },
});
context.displayName = "TableContainer";
const { Provider } = context;

/**
 * This is used to implement the sticky header and footer intersection observer
 * behavior.
 *
 * @internal
 * @remarks \@since 6.0.0
 */
export function useTableContainer(): Readonly<TableContainerContext> {
  return useContext(context);
}

/**
 * @remarks \@since 6.0.0
 */
export function tableContainer({ className }: { className?: string }): string {
  return cnb("rmd-table-container", className);
}

export type TableContainerProps = HTMLAttributes<HTMLDivElement>;

/**
 * An extremely "useful" component that should be used with the `Table`
 * component if you want to make a responsive table within the page. If you
 * don't want to use this component, you can just apply `overflow: auto` to a
 * parent element of the table.
 */
export const TableContainer = forwardRef<HTMLDivElement, TableContainerProps>(
  function TableContainer(props, ref) {
    const { className, children, ...remaining } = props;
    const [nodeRef, refCallback] = useEnsuredRef(ref);

    const value = useMemo(
      () => ({
        exists: true,
        containerRef: nodeRef,
      }),
      [nodeRef]
    );

    return (
      <Provider value={value}>
        <div
          {...remaining}
          ref={refCallback}
          className={tableContainer({ className })}
        >
          {children}
        </div>
      </Provider>
    );
  }
);
