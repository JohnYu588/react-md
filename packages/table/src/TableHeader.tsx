import { bem, useEnsuredRef, useIntersectionObserver } from "@react-md/core";
import { cnb } from "cnbuilder";
import type { HTMLAttributes } from "react";
import { forwardRef, useCallback, useMemo, useState } from "react";

import type {
  TableCellConfig,
  TableConfigContext,
} from "./TableConfigurationProvider";
import {
  TableConfigProvider,
  useTableConfig,
} from "./TableConfigurationProvider";
import { useTableContainer } from "./TableContainer";
import type { TableStickySectionProps } from "./types";

const styles = bem("rmd-thead");

/** @remarks \@since 6.0.0 */
export interface TableHeaderClassNameOptions {
  className?: string;
  dense?: boolean;
  sticky?: boolean;
  stickyActive?: boolean;
}

/** @remarks \@since 6.0.0 */
export function tableHeader(options: TableHeaderClassNameOptions = {}): string {
  const { dense, sticky, stickyActive, className } = options;

  return cnb(
    styles({
      dense,
      sticky,
      "sticky-active": stickyActive,
    }),
    className
  );
}

/**
 * @remarks \@since 6.0.0 Added support for "sticky-active" state.
 */
export interface TableHeaderProps
  extends HTMLAttributes<HTMLTableSectionElement>,
    Pick<TableCellConfig, "lineWrap">,
    TableStickySectionProps {
  /**
   * This is a rename of the `disableHover` of the `TableConfig` since table
   * headers are not hoverable by default. This prop can be enabled to add the
   * row hover color within table headers, but it is not really recommended.
   *
   * @defaultValue `false`
   */
  hoverable?: boolean;
}

/**
 * @remarks \@since 6.0.0
 */
export const isTableHeaderStickyActive = (
  entry: IntersectionObserverEntry
): boolean => {
  return (
    entry.intersectionRatio < 1 &&
    entry.boundingClientRect.bottom <= window.innerHeight
  );
};

/**
 * Creates a `<thead>` element with some basic styles. This component will also
 * update the table configuration so that all the `TableCell` children will
 * automatically become `<th>` elements instead of the normal `<td>` as well as
 * disabling the hover effect and line wrapping. The hover effect and
 * line-wrapping can be re-enabled if desired through the `hoverable` and
 * `disableNoWrap` props.
 */
export const TableHeader = forwardRef<
  HTMLTableSectionElement,
  TableHeaderProps
>(function TableHeader(props, propRef) {
  const {
    className,
    hoverable = false,
    lineWrap: propLineWrap,
    children,
    sticky = false,
    stickyOptions,
    isStickyActive = isTableHeaderStickyActive,
    disableStickyStyles = false,
    ...remaining
  } = props;

  // update the table configuration with the custom overrides for the `<thead>`
  const { dense, hAlign, vAlign, lineWrap, disableHover, disableBorders } =
    useTableConfig({
      lineWrap: propLineWrap,
      disableHover: !hoverable,
    });

  const configuration = useMemo<TableConfigContext>(
    () => ({
      dense,
      header: true,
      hAlign,
      vAlign,
      lineWrap,
      disableBorders,
      disableHover,
    }),
    [dense, hAlign, vAlign, lineWrap, disableBorders, disableHover]
  );

  const [theadRef, theadRefCallback] = useEnsuredRef(propRef);
  const { exists, containerRef } = useTableContainer();
  const [stickyActive, setStickyActive] = useState(false);
  const targetRef = useIntersectionObserver({
    ref: exists ? undefined : theadRefCallback,
    root: containerRef,
    disabled: !sticky || disableStickyStyles,
    threshold: exists ? 0 : 1,
    getRootMargin: useCallback(() => {
      const thead = theadRef.current;
      if (!thead) {
        return;
      }

      let topOffset: number;
      if (exists) {
        topOffset = thead.offsetHeight - 1;
      } else {
        const top = parseFloat(window.getComputedStyle(thead).top);
        topOffset = Number.isNaN(top) ? 1 : top + 1;
      }

      return `-${topOffset}px 0px 0px`;
    }, [exists, theadRef]),
    onUpdate: useCallback(
      (entry) => {
        setStickyActive(isStickyActive(entry));
      },
      [isStickyActive]
    ),
    // allow the user defined sticky options to override the default behavior
    ...stickyOptions,
  });

  return (
    <TableConfigProvider value={configuration}>
      <thead
        {...remaining}
        ref={exists ? theadRefCallback : targetRef}
        className={tableHeader({
          dense,
          sticky,
          stickyActive,
          className,
        })}
      >
        {children}
      </thead>
      {exists && sticky && !disableStickyStyles && (
        // rendering a `<tbody>` since it is valid to have 0-many in a table
        // https://html.spec.whatwg.org/multipage/tables.html#the-table-element
        <tbody aria-hidden ref={targetRef} />
      )}
    </TableConfigProvider>
  );
});
