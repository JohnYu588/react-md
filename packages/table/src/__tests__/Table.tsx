import React, { ReactElement } from "react";
import { render } from "@testing-library/react";

import { Table, TableProps } from "../Table";
import { TableHeader } from "../TableHeader";
import { TableRow } from "../TableRow";
import { TableCell } from "../TableCell";
import { TableBody } from "../TableBody";
import { TableContainer } from "../TableContainer";

function Test(props: TableProps): ReactElement {
  return (
    <Table {...props}>
      <TableHeader>
        <TableRow>
          <TableCell header>Header</TableCell>
        </TableRow>
      </TableHeader>
      <TableBody>
        <TableRow>
          <TableCell>Cell 1</TableCell>
        </TableRow>
      </TableBody>
    </Table>
  );
}

describe("Table", () => {
  it("should render correctly", () => {
    const { container, rerender } = render(<Test />);
    expect(container).toMatchSnapshot();

    rerender(<Test dense hAlign="right" />);
    expect(container).toMatchSnapshot();

    rerender(<Test lineWrap disableHover disableBorders />);
    expect(container).toMatchSnapshot();

    rerender(<Test vAlign="top" fullWidth />);
    expect(container).toMatchSnapshot();

    rerender(
      <TableContainer>
        <Test />
      </TableContainer>
    );
    expect(container).toMatchSnapshot();

    rerender(
      <TableContainer className="test">
        <Test />
      </TableContainer>
    );
    expect(container).toMatchSnapshot();
  });
});
