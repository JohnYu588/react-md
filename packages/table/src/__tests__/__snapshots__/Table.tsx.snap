// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table should render correctly 1`] = `
<div>
  <table
    class="rmd-table"
  >
    <thead
      class="rmd-thead"
    >
      <tr
        class="rmd-tr rmd-tr--bordered"
      >
        <th
          class="rmd-table-cell rmd-table-cell--header rmd-table-cell--no-wrap"
          scope="col"
        >
          Header
        </th>
      </tr>
    </thead>
    <tbody
      class="rmd-tbody"
    >
      <tr
        class="rmd-tr rmd-tr--bordered rmd-tr--hoverable"
      >
        <td
          class="rmd-table-cell rmd-table-cell--no-wrap"
        >
          Cell 1
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`Table should render correctly 2`] = `
<div>
  <table
    class="rmd-table rmd-table--dense"
  >
    <thead
      class="rmd-thead"
    >
      <tr
        class="rmd-tr rmd-tr--bordered"
      >
        <th
          class="rmd-table-cell rmd-table-cell--header rmd-table-cell--right rmd-table-cell--no-wrap"
          scope="col"
        >
          Header
        </th>
      </tr>
    </thead>
    <tbody
      class="rmd-tbody"
    >
      <tr
        class="rmd-tr rmd-tr--bordered rmd-tr--hoverable"
      >
        <td
          class="rmd-table-cell rmd-table-cell--right rmd-table-cell--no-wrap"
        >
          Cell 1
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`Table should render correctly 3`] = `
<div>
  <table
    class="rmd-table"
  >
    <thead
      class="rmd-thead"
    >
      <tr
        class="rmd-tr"
      >
        <th
          class="rmd-table-cell rmd-table-cell--header"
          scope="col"
        >
          Header
        </th>
      </tr>
    </thead>
    <tbody
      class="rmd-tbody"
    >
      <tr
        class="rmd-tr"
      >
        <td
          class="rmd-table-cell"
        >
          Cell 1
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`Table should render correctly 4`] = `
<div>
  <table
    class="rmd-table rmd-table--full-width"
  >
    <thead
      class="rmd-thead"
    >
      <tr
        class="rmd-tr rmd-tr--bordered"
      >
        <th
          class="rmd-table-cell rmd-table-cell--header rmd-table-cell--top rmd-table-cell--vertical rmd-table-cell--no-wrap"
          scope="col"
        >
          Header
        </th>
      </tr>
    </thead>
    <tbody
      class="rmd-tbody"
    >
      <tr
        class="rmd-tr rmd-tr--bordered rmd-tr--hoverable"
      >
        <td
          class="rmd-table-cell rmd-table-cell--top rmd-table-cell--vertical rmd-table-cell--no-wrap"
        >
          Cell 1
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;

exports[`Table should render correctly 5`] = `
<div>
  <div
    class="rmd-table-container"
  >
    <table
      class="rmd-table"
    >
      <thead
        class="rmd-thead"
      >
        <tr
          class="rmd-tr rmd-tr--bordered"
        >
          <th
            class="rmd-table-cell rmd-table-cell--header rmd-table-cell--no-wrap"
            scope="col"
          >
            Header
          </th>
        </tr>
      </thead>
      <tbody
        class="rmd-tbody"
      >
        <tr
          class="rmd-tr rmd-tr--bordered rmd-tr--hoverable"
        >
          <td
            class="rmd-table-cell rmd-table-cell--no-wrap"
          >
            Cell 1
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;

exports[`Table should render correctly 6`] = `
<div>
  <div
    class="rmd-table-container test"
  >
    <table
      class="rmd-table"
    >
      <thead
        class="rmd-thead"
      >
        <tr
          class="rmd-tr rmd-tr--bordered"
        >
          <th
            class="rmd-table-cell rmd-table-cell--header rmd-table-cell--no-wrap"
            scope="col"
          >
            Header
          </th>
        </tr>
      </thead>
      <tbody
        class="rmd-tbody"
      >
        <tr
          class="rmd-tr rmd-tr--bordered rmd-tr--hoverable"
        >
          <td
            class="rmd-table-cell rmd-table-cell--no-wrap"
          >
            Cell 1
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
